staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -178,19 +185,64 @@ func (q *Type) Done(item interface{}) {
 	if q.dirty.has(item) {
 		q.queue = append(q.queue, item)
 		q.cond.Signal()
+	} else if q.processing.len() == 0 {
+		q.cond.Signal()
 	}
 }
 
+// isProcessing indicates if there are still items on the work queue being
+// processed. It's used to drain the work queue on an eventual shutdown.
+func (q *Type) isProcessing() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.processing.len() != 0
+}
+
+// shouldDrain indicates if the queue shutdown should drain.
+func (q *Type) shouldDrain() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.drain
+}
+
 // ShutDown will cause q to ignore all new items added to it. As soon as the
 // worker goroutines have drained the existing items in the queue, they will be
 // instructed to exit.
 func (q *Type) ShutDown() {
+	q.shutdown()
+	for q.isProcessing() && q.shouldDrain() {
```

Comment:
What prevents us from being in a state where we are not processing, but still have items in our queue?

Reply:
Nothing. One example of this is what I mentioned in the description to this PR:

> This PR assumes that whatever is put on the queue will also terminate at one point (I am not sure if this is a valid assumption), i.e: that no syncHandler will indefinitely execute

And if there's a caller (controller) for which that statement does not hold, then we block on `ShutDown` until the second ctrl+c.  

I think for what concerns the logic in `queue.go` : there is no condition I could think of that would lead us to not process and still have items in our queue.